### PR TITLE
Add gradle wrapper to project template

### PR DESCRIPTION
**Pull request by mravn-google**
Introduce use of gradle wrapper in project template to ensure repeatable builds. Manage gradle wrapper scripts and binary using Flutter tool cache. Update gradle and android gradle plugin versions to address #9842 and #10236.

Notes:
- Upgrades new Flutter projects to use gradle wrapper + gradle version 3.3 + android gradle plugin version 2.3.3. This is the configuration used by Android Studio in latest stable version 2.3.3.
- PR is morally a retake of #9882, but without distributing the gradle wrapper scripts and binaries using the Flutter repo. Instead, these files are managed by the cache, similar to the engine dependency.
- `cache.dart` has been refactored to avoid duplication across now three cached artifacts (material fonts, engine, gradle wrapper).
- There should be no implications for existing Flutter projects.
- New projects will get the [standard complement](https://docs.gradle.org/3.3/userguide/gradle_wrapper.html#sec:wrapper_generation) of gradle wrapper files in their `android` folder. These files should be checked into version control.
- #10927 is left for a separate task.

### CODE REVIEW

**Comment by mravn-google** [comment]
cc @abarth @Hixie

**Review by mit-mit — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/cache.dart [hunk 1]
```diff
@@ -188,77 +191,63 @@ class Cache {
...
+    return isUpToDateInner();
   }
 
-  Future<Null> download() {
-    final Status status = logger.startProgress('Downloading Material fonts...', expectSlowOperation: true);
+  Future<Null> update() async {
+    if (location.existsSync())
+      location.deleteSync(recursive: true);
+    location.createSync(recursive: true);
+    return updateInner().then<Null>((Null value) {
```
Comment (jakobr-google) [inline packages/flutter_tools/lib/src/cache.dart]: `_` instead of `Null value`?

**Review by jakobr-google — APPROVED** [review]
LGTM

File: bin/internal/gradle_wrapper.version [hunk 2]
```diff
@@ -0,0 +1 @@
+0b5c1398d1d04ac245a310de98825cb7b3278e2a
```
Comment (mravn-google) [inline bin/internal/gradle_wrapper.version]: That could trip up parsing. The version files for engine and material fonts do not include such a comment; file name says it all I guess.

File: packages/flutter_tools/lib/src/cache.dart (hunk 1 above)
Comment (mravn-google) [inline packages/flutter_tools/lib/src/cache.dart]: Done.

**Pull request merged**